File: src/main/java/insogroup/HtmlParser/StandardRealisation/PluginSyntaxAnalyzer/PluginSaxParser.java
package insogroup.HtmlParser.StandardRealisation.PluginSyntaxAnalyzer;

import insogroup.HtmlParser.StandardRealisation.exceptions.PluginException;
import insogroup.HtmlParser.StandardRealisation.settings.classes.Settings;
import org.xml.sax.Attributes;
import org.xml.sax.helpers.DefaultHandler;

public class PluginSaxParser extends DefaultHandler{

    private String spaces;
    private String tab;

    private Settings settings;

    private State currentState;

    public PluginSaxParser(Settings settings) {
        this.settings = settings;
        spaces = "";
        tab = "    ";
    }

    @Override
    public void startDocument(){
    }

    @Override
    public void startElement(String uri, String localName, String qName, Attributes attr){
        spaces += tab;
        try {
            currentState = currentState.addChild(qName, attr);
        }
        catch (NullPointerException e){
            currentState = new State(null, qName, attr, settings);
        }
    }

    @Override
    public void characters(char[] text, int start, int length){
        String stringText = "";
        for (int i = start; i< start + length; i++)
            stringText += text[i];
        currentState.setText(stringText);
    }

    @Override
    public void endElement(String uri, String localName, String qName){
        spaces = spaces.replaceFirst(tab, "");
        try{
            currentState = currentState.getParent();
        }
        catch (PluginException e){}
    }

    @Override
    public void endDocument(){
//        mediator.print("End parsing plugin, thank you.");
    }

    public State getRootState(){
        State res = currentState;
        while (true){
            try{
                res = res.getParent();
            }
            catch (PluginException e){
                break;
            }
        }
        return res;
    }

}

File: src/main/kotlin/com/github/insanusmokrassar/HTMLParser/StandardRealisation/PluginSyntaxAnalyzer/PluginSaxParser.kt
package com.github.insanusmokrassar.HTMLParser.StandardRealisation.PluginSyntaxAnalyzer

import com.github.insanusmokrassar.HTMLParser.StandardRealisation.PluginSyntaxAnalyzer.interfaces.State
import com.github.insanusmokrassar.HTMLParser.StandardRealisation.Settings
import com.github.insanusmokrassar.HTMLParser.StandardRealisation.exceptions.PluginException
import org.xml.sax.Attributes
import org.xml.sax.helpers.DefaultHandler

class PluginSaxParser(private val settings: Settings) : DefaultHandler() {

    private var spaces: String? = null
    private val tab: String

    private var currentState: State? = null

    val rootState: State?
        get() = currentState ?. root

    init {
        spaces = ""
        tab = "    "
    }

    override fun startElement(uri: String, localName: String?, qName: String, attr: Attributes) {
        spaces += tab
        currentState = currentState ?. addChild(qName, attr) ?: PluginState(null, qName, attr, settings)
    }

    override fun characters(text: CharArray?, start: Int, length: Int) {
        var stringText = ""
        for (i in start until start + length)
            stringText += text!![i]
        currentState ?. text = stringText
    }

    override fun endElement(uri: String?, localName: String?, qName: String?) {
        spaces = spaces!!.replaceFirst(tab.toRegex(), "")
        try {
            currentState ?. parent ?.let {
                currentState = it
            }
        } catch (e: PluginException) {
        }
    }
}
